builtin/logical/mongodb/backend.go

```diff
@@ -0,0 +1,122 @@
+package mongodb
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"gopkg.in/mgo.v2"
+)
+
+func Factory(conf *logical.BackendConfig) (logical.Backend, error) {
+	return Backend().Setup(conf)
+}
+
+func Backend() *framework.Backend {
+	var b backend
+	b.Backend = &framework.Backend{
+		Help: strings.TrimSpace(backendHelp),
+
+		Paths: []*framework.Path{
+			pathConfigConnection(&b),
+			pathConfigLease(&b),
+			pathListRoles(&b),
+			pathRoles(&b),
+			pathCredsCreate(&b),
+		},
+
+		Secrets: []*framework.Secret{
+			secretCreds(&b),
+		},
+
+		Clean: b.ResetSession,
+	}
+
+	return b.Backend
+}
+
+type backend struct {
+	*framework.Backend
+
+	session *mgo.Session
+	lock    sync.Mutex
+}
+
+// Session returns the database connection.
+func (b *backend) Session(s logical.Storage) (*mgo.Session, error) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// If we already have a Session, we got it!
+	if b.session != nil {
```

Comment:
Can you `Ping()` to ensure that the connection is still valid, and close/reopen if not? See https://github.com/hashicorp/vault/blob/master/builtin/logical/postgresql/backend.go#L59